crypto/kem/README.md

```diff
@@ -0,0 +1,227 @@
+                                                                                  encaps_ctx = EVP_PKEY_CTX_new(key, NULL);
+                                                                                  if (!encaps_ctx) {
+                                                                                    goto end;
+                                                                                  }
+
+                                                                                  // Get the required buffer lengths and allocate the buffers.
+                                                                                  if (!EVP_PKEY_encapsulate(encaps_ctx, NULL, &ct_len, NULL, &ss_len)) {
+                                                                                    goto end;
+                                                                                  }
+                                                                                  ct = (uint8_t*) OPENSSL_malloc(ct_len);
+                                                                                  ss = (uint8_t*) OPENSSL_malloc(ss_len);
+                                                                                  if (!ct || !ss) {
+                                                                                    goto end;
+                                                                                  }
+
+                                                                                  // Encapsulate.
+                                                                                  if (!EVP_PKEY_encapsulate(encaps_ctx, ct, &ct_len, ss, &ss_len)) {
+                                                                                    goto end;
+                                                                                  }
+
+                                                                                  ret = SUCCESS;
+
+                                                                                  end:
+
+                                                                                  OPENSSL_free(ss);
+                                                                                  OPENSSL_free(ct);
+                                                                                  EVP_PKEY_free(key);
+                                                                                  EVP_PKEY_CTX_free(encaps_ctx);
+
+                                                                                  // We don't free the shared secret buffer because the whole point
+                                                                                  // of the protocol is to generate it, so we consider it as the output.
+
+                                                                                  return ret;
+                                                                    ct                
```

Comment:
I'm confused a bit whether `ct` is freed or not because it's sent. And what does the comment refer to as the `shared secret buffer`?

Reply:
right, I don't know exactly how customers are going to implement the protocol in terms of sending, waiting for response, etc. Anyway, I'll update the example to show three separate full functions. Hopefully, it'll be more clear then